### ISSUE TITLE
Rule proposal: `no-immediate-mutation`

### ISSUE THREAD

**Issue by fisker** [sindresorhus/eslint-plugin-unicorn#2775]
### Description

Forbid immediate array mutation after declaration.

### Examples

```js
// ❌
const array = [1, 2];
array.push(3, 4);

// ✅
const array = [1, 2, 3, 4];
```

```js

// ❌
const array = [3, 4];
array.unshift(1, 2);

// ✅
const array = [1, 2, 3, 4];
```

### Proposed rule name

no-immediate-array-mutation

### Additional Info

Some cases are hard to detect.

```js
const array = [1, 2];
array.push(foo()); // <-- this `foo` funtion may have access to the original `array`.
```

**Comment by silverwind** [comment]
> Do we have more cases to detect, except the Set case?

Yes, `Set`, `WeakSet`, `Map`and `WeakMap` could be handled when they are initialized from an array literal and it should flag all immediately following `add` calls.

> Should we detect the following cases for object?

Not sure. It seems with `Object.assign`, the autor's intention is not as clear as in the other cases.

**Comment by fisker** [comment]
@sindresorhus What's your opinion on the `Object.assign()` case? https://github.com/sindresorhus/eslint-plugin-unicorn/issues/2775#issuecomment-3434304022

**Comment by sindresorhus** [comment]
Yes, makes sense to me.

**Comment by fisker** [comment]
I think this rule make sense to apply to assignments too.

```
// ❌
let array = []

// Do something with array

array = [1, 2];
array.push(3, 4);

// ✅
let array = []

// Do something with array

array = [1, 2, 3, 4];
```

This rule will be huge!
